### ISSUE TITLE
Move creation of Lucene document to ItemProcessor

### ISSUE THREAD

**Issue by gunnarmorling** [mincong-h/gsoc-hsearch#102]

**Comment by mincong-h** [comment]
The creation of Lucene document is now moved to ItemProcessor.

When the number of entities is small (e.g. 100), the job works fine. When the number of entities is bigger (e.g. 1000), then there's something weird : 

* The TX timeout is reached at 444 works done. Warning with ARJUNA012117.
* The job processed all the works successfully, but still became failed at 1000 works done. Error with ARJUNA016102.

I've tried to increase the TX timeout in the job XML file : 

```xml
<step id="produceLuceneDoc">
    <properties>
        <property name="javax.transaction.global.mode" value="true" />
        <property name="javax.transaction.global.timeout" value="300" /><!-- seconds -->
    </properties>
    <chunk checkpoint-policy="custom">
```

But it didn't help. Notice that the default timeout is 180 seconds, which is already more than enough.

## Log

Here's the full log for both actions : 

> 00:12:52,423 INFO  [org.hibernate.search.jsr352.internal.steps.lucene.ItemWriter] (Batch Thread - 2) checkpointInfo called
00:12:52,423 WARN  [com.arjuna.ats.arjuna] (Transaction Reaper) ARJUNA012117: TransactionReaper::check timeout for TX 0:ffffc0a8010c:1e39df7e:5786bcd5:bcb in state  RUN
00:12:52,425 INFO  [org.hibernate.search.jsr352.internal.steps.lucene.PartitionAnalyzer] (Batch Thread - 1) 444 works done
00:12:52,426 WARN  [com.arjuna.ats.arjuna] (Transaction Reaper Worker 0) ARJUNA012095: Abort of action id 0:ffffc0a8010c:1e39df7e:5786bcd5:bcb invoked while multiple threads active within it.
00:12:52,427 WARN  [com.arjuna.ats.arjuna] (Transaction Reaper Worker 0) ARJUNA012381: Action id 0:ffffc0a8010c:1e39df7e:5786bcd5:bcb completed with multiple threads - thread Batch Thread - 1 was in progress with sun.misc.Unsafe.park(Native Method)
java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
java.util.concurrent.LinkedBlockingQueue.take(LinkedBlockingQueue.java:442)
org.jberet.runtime.runner.StepExecutionRunner.beginPartition(StepExecutionRunner.java:350)
org.jberet.runtime.runner.StepExecutionRunner.runBatchletOrChunk(StepExecutionRunner.java:216)
org.jberet.runtime.runner.StepExecutionRunner.run(StepExecutionRunner.java:140)
org.jberet.runtime.runner.CompositeExecutionRunner.runStep(CompositeExecutionRunner.java:164)
org.jberet.runtime.runner.CompositeExecutionRunner.runFromHeadOrRestartPoint(CompositeExecutionRunner.java:88)
org.jberet.runtime.runner.JobExecutionRunner.run(JobExecutionRunner.java:59)
org.wildfly.extension.batch.jberet.impl.BatchEnvironmentService$WildFlyBatchEnvironment$1.run(BatchEnvironmentService.java:243)
org.wildfly.extension.requestcontroller.RequestController$QueuedTask$1.run(RequestController.java:497)
org.jberet.spi.JobExecutor$3.run(JobExecutor.java:161)
org.jberet.spi.JobExecutor$1.run(JobExecutor.java:99)
java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
java.lang.Thread.run(Thread.java:745)
org.jboss.threads.JBossThread.run(JBossThread.java:320)

> 00:12:52,427 WARN  [com.arjuna.ats.arjuna] (Transaction Reaper Worker 0) ARJUNA012108: CheckedAction::check - atomic action 0:ffffc0a8010c:1e39df7e:5786bcd5:bcb aborting with 1 threads active!
00:12:52,428 WARN  [com.arjuna.ats.arjuna] (Transaction Reaper Worker 0) ARJUNA012121: TransactionReaper::doCancellations worker Thread[Transaction Reaper Worker 0,5,main] successfully canceled TX 0:ffffc0a8010c:1e39df7e:5786bcd5:bcb

---

> 00:12:54,103 ERROR [org.jberet] (Batch Thread - 1) JBERET000007: Failed to run job massIndex, produceLuceneDoc, org.jberet.job.model.Step@7eb78df9: javax.transaction.RollbackException: ARJUNA016102: The transaction is not active! Uid is 0:ffffc0a8010c:1e39df7e:5786bcd5:bcb
	at com.arjuna.ats.internal.jta.transaction.arjunacore.TransactionImple.commitAndDisassociate(TransactionImple.java:1190)
	at com.arjuna.ats.internal.jta.transaction.arjunacore.BaseTransaction.commit(BaseTransaction.java:126)
	at com.arjuna.ats.jbossatx.BaseTransactionManagerDelegate.commit(BaseTransactionManagerDelegate.java:89)
	at org.jberet.runtime.runner.StepExecutionRunner.beginPartition(StepExecutionRunner.java:388)
	at org.jberet.runtime.runner.StepExecutionRunner.runBatchletOrChunk(StepExecutionRunner.java:216)
	at org.jberet.runtime.runner.StepExecutionRunner.run(StepExecutionRunner.java:140)
	at org.jberet.runtime.runner.CompositeExecutionRunner.runStep(CompositeExecutionRunner.java:164)
	at org.jberet.runtime.runner.CompositeExecutionRunner.runFromHeadOrRestartPoint(CompositeExecutionRunner.java:88)
	at org.jberet.runtime.runner.JobExecutionRunner.run(JobExecutionRunner.java:59)
	at org.wildfly.extension.batch.jberet.impl.BatchEnvironmentService$WildFlyBatchEnvironment$1.run(BatchEnvironmentService.java:243)
	at org.wildfly.extension.requestcontroller.RequestController$QueuedTask$1.run(RequestController.java:497)
	at org.jberet.spi.JobExecutor$3.run(JobExecutor.java:161)
	at org.jberet.spi.JobExecutor$1.run(JobExecutor.java:99)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
	at org.jboss.threads.JBossThread.run(JBossThread.java:320)

**Comment by mincong-h** [comment]
Fixed in issue #67 :

> The transaction timeout is defined in the WildFly config `standalone-full-testqueues.xml`. Change in from 2 seconds to 300 seconds and everything works fine.

So every component works fine now and creation of Lucene document has been moved to ItemProcessor. I'll increase the test dataset from 100 to 1000, then close this issue.
